JP-3697: Jump Step Refactor

kmacdonald-stsci opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->

Resolves [JP-3697](https://jira.stsci.edu/browse/JP-3697)

<!-- If this PR closes a GitHub issue, reference it here by its number -->

Closes #

<!-- describe the changes comprising this PR here -->

This PR refactors STCAL focusing on readability, memory consumption, and limiting scope of variables.  This refactor reduced jump step memory reduction.  The resident size of memory is reduced by more than 2GB, with the memory consumption spike due to `numpy.nanmedian` completely eliminated, and the heap size reduced by 1GB.

The interface has been simplified with classes.  Large functions have been logically broken into smaller functions for readability and scope limiting.  Code duplications have been eliminated.  Dead code and duplicated variable definitions have been eliminated.

The `twopoint_difference.py` still needs to be refactored.

All CI tests pass.

All CI tests in the corresponding JWST PR pass, as do all '-k jump' regression tests.

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [x] update or add relevant tests
- [x] update relevant docstrings and / or `docs/` page
- [x] Does this PR change any API used downstream? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types)
  - [ ] run regression tests with this branch installed (`"git+https://github.com/<fork>/stcal@<branch>"`)
    - [ ] [`jwst` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) 
    - [x] [`romancal` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/romancal.yml)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.apichange.rst``: change to public API
- ``changes/<PR#>.bugfix.rst``: fixes an issue
- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
</details>

kmacdonald-stsci commented:
I need to update `test_twopoint_difference.py`.  I'll change it's name to `test_jump_twopoint_difference.py`, so no future developer updating the jump step will miss these tests.

t-brandt commented on src/stcal/jump/jump.py:
```diff
@@ -1118,12 +908,319 @@ def find_faint_extended(
+    data : ndarray
+        NaN'd cience data 4D float
+    """
+    jump_dnu_flag = jump_data.fl_jump + jump_data.fl_dnu
+    sat_dnu_flag = jump_data.fl_sat + jump_data.fl_dnu
+    data[gdq == jump_dnu_flag] = np.nan
+    data[gdq == sat_dnu_flag] = np.nan
+    data[gdq == jump_data.fl_sat] = np.nan
+    data[gdq == jump_data.fl_jump] = np.nan
+    data[gdq == jump_data.fl_dnu] = np.nan
```
```suggestion
    dnu_jump_sat = jump_data.fl_jump | jump_data.fl_dnu | jump_data.fl_sat
    data[gdq & dnu_jump_sat != 0] = np.nan
```
This function is now so limited that I suggest removing it and moving these two lines to where the function is called.

kmacdonald-stsci replied:
Your recommended changes are not logically equivalent to the original code.  For example `0x03==0x02` is false, but `0x03 & 0x02` is true.  The original code isn't looking for any pixel where `dnu`, `jump`, and `sat` are some of the flags are set, but where those flags are the only flags set.

t-brandt replied:
You're right.  It does seem strange to me that this would be the intended behavior.  If these flags are set, but also a hot pixel flag, then we don't want to set the value to nan?

t-brandt commented on src/stcal/jump/jump.py:
```diff
@@ -689,50 +549,32 @@ def extend_ellipses(
+            saty, satx = np.where(sat_pix == jump_data.fl_sat)
             jump_ellipse[saty, satx] = 0
             out_gdq_cube[intg, flg_grp, :, :] = np.bitwise_or(gdq_cube[intg, flg_grp, :, :], jump_ellipse)
     diff_cube = out_gdq_cube - gdq_cube
```
```suggestion
```
Expensive, unused calculation.

kmacdonald-stsci replied:
Removed.

t-brandt replied:
Looks like it's still there.  This line can actually double the runtime of NIR data with many integrations.

kmacdonald-stsci replied:
I thought I removed it before.  Let's see if I can get it right this time.

t-brandt commented on src/stcal/jump/jump.py:
```diff
@@ -689,50 +549,32 @@ def extend_ellipses(
+            sat_pix = np.bitwise_and(gdq_cube[intg, flg_grp, :, :], jump_data.fl_sat)
+            saty, satx = np.where(sat_pix == jump_data.fl_sat)
             jump_ellipse[saty, satx] = 0
             out_gdq_cube[intg, flg_grp, :, :] = np.bitwise_or(gdq_cube[intg, flg_grp, :, :], jump_ellipse)
```
This line is repeating a lot of calculations at great (and unneeded) expense.  Can wait for my rewrite/refactor of some of these functions to fix it, or we can fix all of it here.  If we fix it here there will be somewhat more extensive changes.

melanieclarke replied:
Let's defer any performance-related changes that are more extensive than single line fixes and get the general refactor in first.  That should give us a more solid base to start from

t-brandt commented on src/stcal/jump/jump.py:
```diff
@@ -846,221 +738,92 @@ def near_edge(jump, low_threshold, high_threshold):
 
     all_ellipses = []
 
     first_diffs_masked = np.ma.masked_array(first_diffs, mask=np.isnan(first_diffs))
```
masked_array not needed since every usage below uses nanmedian or otherwise ignores nan when collapsing along any axis.  masked_array slows things and adds memory and overhead.

kmacdonald-stsci replied:
The `masked_array` is needed to return the proper class type to be used in the computation of `ratio`.  The `ratio` instance needs the `filled` method, which isn't present in the `first_diffs` instance.  Not using `masked_array` here results in a failure here:

https://github.com/spacetelescope/stcal/blob/5759c19a0ce7cfe66d6230f77c078c84ab516875/src/stcal/jump/jump.py#L1074

t-brandt replied:
Strictly true, but since the mask is on nan pixels anyway, the filled method is also redundant and unneeded.  We can defer all of this to performance improvements later.  There is a lot of redundant and/or unneeded work being done in the base branch.

kmacdonald-stsci commented on src/stcal/jump/jump.py:
```diff
@@ -81,423 +37,318 @@ def detect_jumps(
-        flagged.
+    twopt_params = TwoPointParams(jump_data, False)
+    if n_slices == 1:
+        twopt_params.minimum_groups = 3  # XXX Should this be hard coded as 3?
```
I think it might be better to check to make sure this value is at least 3, so it can be a higher number if desired.  But I didn't want to make this change unless this was intentionally hard coded as 3.

melanieclarke replied:
Are we deferring this discussion for future work?

t-brandt replied:
I vote to defer discussion.  I think this is fine for now.

kmacdonald-stsci commented on src/stcal/jump/jump.py:
```diff
@@ -81,423 +37,318 @@ def detect_jumps(
-        not have its neighbors flagged.
+    # remove redundant bits in pixels that have jump flagged but were
+    # already flagged as do_not_use or saturated.
+    gdq[gdq == np.bitwise_or(dnu, jump)] = dnu
+    gdq[gdq == np.bitwise_or(sat, jump)] = sat
```
Below is a link to the pre-refactored code in the current main branch.  The only refactor I did is to remove the usage of the `dqflags` dictionary and use the locally defined flags.  If this isn't correct here, it's not correct in the main branch.

https://github.com/spacetelescope/stcal/blob/d6a87c346b56daafcd6b8b43231fdbb8b9222965/src/stcal/jump/jump.py#L298

melanieclarke replied:
I agree with Tim - this looks incorrect, in both the main and refactor branches.  I don't think this needs a whole separate ticket, so let's go ahead and fix the bitwise logic here.

t-brandt replied:
Looking at this again, it will correctly unset the jump flag if only dnu and jump are set, or if only sat and jump are set.  If this is the intended behavior then I think it is fine.  If the intended behavior is to unset the jump flag if either dnu or sat are set, then the lines may not do that.

kmacdonald-stsci replied:
This change causes an error in `tests/test_jump.py::test_multiprocessing`, but passes all JWST CI and regression tests.  I will debug this failure before committing.

melanieclarke replied:
I think the XOR logic is backward in your commented out code - we want to remove jump, not dnu, so `gdq[gdq == np.bitwise_or(dnu, jump)] ^= jump`

kmacdonald-stsci replied:
```
gdq[gdq == np.bitwise_or(dnu, jump)] = dnu
gdq[gdq == np.bitwise_or(sat, jump)] = sat

```
The above is logically equivalent to:

```
dnu_jump = dnu | jump
sat_jump = sat | jump
gdq[gdq == dnu_jump] = dnu
gdq[gdq == sat_jump] = sat
```

But I think it should be (but this fails tests):

```
dnu_jump = dnu | jump
sat_jump = sat | jump
gdq[np.bitwise_and(gdq, dnu_jump).astype(bool)] ^= jump
gdq[np.bitwise_and(gdq, sat_jump).astype(bool)] ^= jump
```

melanieclarke replied:
I agree, I think that logic needs to change - we should remove jump wherever dnu and sat are already set, even if there are other informational flags. Let me take a look at the test.

kmacdonald-stsci replied:
I think @t-brandt got the logic correct above.  I'll implement what he has above and test it.

melanieclarke replied:
The issue is that `np.bitwise_and(gdq, dnu_jump).astype(bool)` matches both jump and dnu_jump, so it's removing lone jump flags.  I think this would work:
```
    gdq[(gdq & dnu_jump) == dnu_jump] -= jump
    gdq[(gdq & sat_jump) == sat_jump] -= jump
```

kmacdonald-stsci commented on src/stcal/jump/jump.py:
```diff
@@ -81,423 +37,318 @@ def detect_jumps(
+            # row of the previous slice and flag any neighbors in the
+            # bottom row of this slice saved from the top of the previous
+            # slice
+            gdq[:, :, k * yinc - 1, :] = np.bitwise_or(gdq[:, :, k * yinc - 1, :], row_below_gdq[:, :, :])
+            gdq[:, :, k * yinc, :] = np.bitwise_or(
+                gdq[:, :, k * yinc, :], previous_row_above_gdq[:, :, :]
+            )
```
The refactor here is the creation of the function `reassemble_sliced_data`.  I did not refactor this part of the code, which is here in the current main branch:

https://github.com/spacetelescope/stcal/blob/d6a87c346b56daafcd6b8b43231fdbb8b9222965/src/stcal/jump/jump.py#L415

melanieclarke replied:
Let's go ahead and clean this up too.

kmacdonald-stsci commented on src/stcal/jump/jump.py:
```diff
@@ -81,423 +37,318 @@ def detect_jumps(
+        says how many rows to increment to get to the next slice.
+    """
+    nrows = data.shape[2]
+    yinc = int(nrows // n_slices)
```
Removed `int` function.

kmacdonald-stsci commented on src/stcal/jump/jump.py:
```diff
@@ -81,423 +37,318 @@ def detect_jumps(
+        The pixel DQ array (2D)
+    """
+    pdq = jump_data.pdq
+    wh_g = np.where(jump_data.gain_2d <= 0.0)
+    if len(wh_g[0] > 0):
+        pdq[wh_g] = np.bitwise_or(pdq[wh_g], jump_data.fl_ngv)
+        pdq[wh_g] = np.bitwise_or(pdq[wh_g], jump_data.fl_dnu)
+
+    wh_g = np.where(np.isnan(jump_data.gain_2d))
+    if len(wh_g[0] > 0):
+        pdq[wh_g] = np.bitwise_or(pdq[wh_g], jump_data.fl_ngv)
+        pdq[wh_g] = np.bitwise_or(pdq[wh_g], jump_data.fl_dnu)
```
I changed the function to remove the `np.where`, but I favor using the function.

kmacdonald-stsci commented on src/stcal/jump/jump.py:
```diff
@@ -510,173 +361,182 @@ def flag_large_events(
-                (0, 0, 22),  # in the RGB cube, set blue plane pixels of the ellipse to 22
-                -1,
-            )
+            color = (0, 0, 22)  # in the RGB cube, set blue plane pixels of the ellipse to 22
```
Changed.

melanieclarke reviewed (commented):
I have not reviewed extensively yet, but so far this looks like a significant improvement in organization.

One thing that needs general clean up -- many of the new docstrings (thanks for adding them!) have some format compliance issues.  In general, the sections need to be named "Parameters" and "Returns" -- some are named "Parameter" and "Return", or even "Paramter" -- and need to have dashes below them, matching the length of the string (not underscores).  The one-line summary should be in imperative case and end in a period (e.g. "Do thing." rather than "This does thing").

To help catch these things, try this from the top level of the stcal repo:
`ruff check src/stcal/jump --select=D`

kmacdonald-stsci commented on src/stcal/jump/jump.py:
```diff
@@ -81,423 +37,318 @@ def detect_jumps(
-                minimum_sigclip_groups,
-                only_use_ints,
-                min_diffs_single_pass,
+                data[:, :, i * yinc: (i + 1) * yinc, :],
+                gdq[:, :, i * yinc: (i + 1) * yinc, :].copy(),
```
Removed this copy.

kmacdonald-stsci commented on src/stcal/jump/jump.py:
```diff
@@ -1118,12 +908,319 @@ def find_faint_extended(
+    masked_ratio[saty, satx] = np.nan
+
+    #  mask pixels that are already flagged as do not use
+    dnu_pixels_array = np.bitwise_and(combined_pixel_mask, 1)
+    dnuy, dnux = np.where(dnu_pixels_array == 1)
+    masked_ratio[dnuy, dnux] = np.nan
```
I changed the `1` to `dnu_flag`.  The computed `dnuy, dnux` is used twice, so kept this `np.where` for the `masked_ratio` and the 'masked_smoothed_ratio`.

kmacdonald-stsci commented on src/stcal/jump/jump_class.py:
```diff
@@ -0,0 +1,569 @@
+        """
+        Set arrays from a data model.
+
+        jump_model : datamodel
```
I believe I changed all these.  I also expanded some docstring comments.

melanieclarke commented on changes/330.jump.rst:
```diff
@@ -0,0 +1,2 @@
+Refactoring the jump module to reduce memory consumption, readability,
+and maintenance.
```
Minor suggestion for rewording here - we're definitely not reducing readability here!

Also, I think the compiled version of these change notes looks better if the note is all on one line - otherwise it adds unnecessary breaks.
```suggestion
Refactoring the jump module to improve memory consumption, readability, and maintenance.
```

melanieclarke commented on src/stcal/jump/jump.py:
```diff
@@ -81,426 +41,311 @@ def detect_jumps(
+    gdq : ndarray
+        The group DQ, 4D array uint8.
+
+    total_primary_crs
```
This return value is missing documentation

melanieclarke commented on src/stcal/jump/jump.py:
```diff
@@ -806,38 +668,67 @@ def make_snowballs(
+    Jumps that are within the threshold will not require a saturated core
+    since this may be off the detector
+
+    Parameter
+    ---------
```
This one still needs fixing

melanieclarke commented on src/stcal/jump/jump.py:
```diff
@@ -1070,27 +832,54 @@ def find_faint_extended(
+    """
+    Ensure that flagging showers didn't change final fluxes by more than allowed.
+
+    Parameter
+    ---------
```
This one still needs fixing

ddavis-stsci commented:
The romancal regression tests pass with this PR,
https://github.com/spacetelescope/RegressionTests/actions/runs/12795655362



On 1/15/25 9:58 AM, Melanie Clarke wrote:
>
> ***@***.**** commented on this pull request.
>
> A few more minor follow up comments, but I think we're almost done 
> here, if @t-brandt 
> <https://urldefense.com/v3/__https://github.com/t-brandt__;!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!0PyMFDr4GDddAO724VMX8HlYagTH3i9YT9Kglbgdn7JjRQIt7q6w2sHg_D7K7kBSioaGbFW7RHzPRgAozAVK2n31$> 
> is happy.
>
> I would like to see a full regtest run for jwst passing before I 
> approve, though, and we should run romancal as well, just in case.
>
> ------------------------------------------------------------------------
>
> In changes/330.jump.rst 
> <https://urldefense.com/v3/__https://github.com/spacetelescope/stcal/pull/330*discussion_r1916785078__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!0PyMFDr4GDddAO724VMX8HlYagTH3i9YT9Kglbgdn7JjRQIt7q6w2sHg_D7K7kBSioaGbFW7RHzPRgAozMaASyUH$>:
>
> > +Refactoring the jump module to reduce memory consumption, readability,
> +and maintenance.
>
> Minor suggestion for rewording here - we're definitely not reducing 
> readability here!
>
> Also, I think the compiled version of these change notes looks better 
> if the note is all on one line - otherwise it adds unnecessary breaks.
>
> ⬇️ Suggested change
> -Refactoring the jump module to reduce memory consumption, readability,
> -and maintenance.
> +Refactoring the jump module to improve memory consumption, readability, and maintenance.
> ------------------------------------------------------------------------
>
> In src/stcal/jump/jump.py 
> <https://urldefense.com/v3/__https://github.com/spacetelescope/stcal/pull/330*discussion_r1916789667__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!0PyMFDr4GDddAO724VMX8HlYagTH3i9YT9Kglbgdn7JjRQIt7q6w2sHg_D7K7kBSioaGbFW7RHzPRgAozCVMu9qW$>:
>
> >  
>       Returns
>       -------
> -    gdq : int, 4D array
> -        updated group dq array
> +    gdq : ndarray
> +        The group DQ, 4D array uint8.
> +
> +    total_primary_crs
>
> This return value is missing documentation
>
> ------------------------------------------------------------------------
>
> In src/stcal/jump/jump.py 
> <https://urldefense.com/v3/__https://github.com/spacetelescope/stcal/pull/330*discussion_r1916793964__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!0PyMFDr4GDddAO724VMX8HlYagTH3i9YT9Kglbgdn7JjRQIt7q6w2sHg_D7K7kBSioaGbFW7RHzPRgAozA7DAETp$>:
>
> >  
> -    max_jump_to_flag_neighbors : float
> -        value in units of sigma that sets the upper limit for flagging of
> -        neighbors. Any jump above this cutoff will not have its neighbors
> -        flagged.
> +    twopt_params = TwoPointParams(jump_data, False)
> +    if n_slices == 1:
> +        twopt_params.minimum_groups = 3  # XXX Should this be hard coded as 3?
>
> Are we deferring this discussion for future work?
>
> ------------------------------------------------------------------------
>
> In src/stcal/jump/jump.py 
> <https://urldefense.com/v3/__https://github.com/spacetelescope/stcal/pull/330*discussion_r1916796235__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!0PyMFDr4GDddAO724VMX8HlYagTH3i9YT9Kglbgdn7JjRQIt7q6w2sHg_D7K7kBSioaGbFW7RHzPRgAozD9i03fx$>:
>
> > +    Parameter
> +    ---------
>
> This one still needs fixing
>
> ------------------------------------------------------------------------
>
> In src/stcal/jump/jump.py 
> <https://urldefense.com/v3/__https://github.com/spacetelescope/stcal/pull/330*discussion_r1916797316__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!0PyMFDr4GDddAO724VMX8HlYagTH3i9YT9Kglbgdn7JjRQIt7q6w2sHg_D7K7kBSioaGbFW7RHzPRgAozMgo7fIS$>:
>
> > +    Parameter
> +    ---------
>
> This one still needs fixing
>
> —
> Reply to this email directly, view it on GitHub 
> <https://urldefense.com/v3/__https://github.com/spacetelescope/stcal/pull/330*pullrequestreview-2552975750__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!0PyMFDr4GDddAO724VMX8HlYagTH3i9YT9Kglbgdn7JjRQIt7q6w2sHg_D7K7kBSioaGbFW7RHzPRgAozELP-kO1$>, 
> or unsubscribe 
> <https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/ALXCXWJZRG3LQQEAJ7BMKWL2KZZPXAVCNFSM6AAAAABUN3KVZOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKNJSHE3TKNZVGA__;!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!0PyMFDr4GDddAO724VMX8HlYagTH3i9YT9Kglbgdn7JjRQIt7q6w2sHg_D7K7kBSioaGbFW7RHzPRgAozCcZzeeC$>.
> You are receiving this because your review was requested.Message ID: 
> ***@***.***>
>

t-brandt commented on src/stcal/jump/jump.py:
```diff
@@ -806,38 +687,70 @@ def make_snowballs(
+
+    Returns
+    -------
+    Boolean if ellipse is close to the detector's edge.
```
```suggestion
    Boolean : True if ellipse is close to the detector's edge.
```

melanieclarke commented:
To recap current status for this PR:
- Ken is currently testing to make sure that the refactor here, with the new JumpClass, is not making memory consumption worse for this step.
- The change note filename needs to be updated: 'jump' is not a valid type.  It probably needs to be 'apichange'.
- I think all other clean up issues have been addressed here.  There are still some remaining clean up issues on the [jwst PR](https://github.com/spacetelescope/jwst/pull/9039) that need to be addressed before merging.
- When that is done, full regression tests need to be run for jwst.
- We will defer further performance improvements for future PRs.

kmacdonald-stsci commented:
> * jwst PR

1. I have done a lot of analysis on the memory with various changes and it appears the increased memory consumption isn't due to the refactor.  The memory consumption once it gets inside the `find_crs` is roughly the same as it is in the main branch, but for some reason when it gets to 

https://github.com/kmacdonald-stsci/stcal/blob/777e772bc3892ac06d60a0398d7bd92f9be29ff6/src/stcal/jump/twopoint_difference.py#L328

the memory consumption is greater in the refactored code than in the main; however, when a lot of this memory gets released, the heap size is smaller in the refactor than in the main, despite peak memory consumption being higher in the refactor.  I think Tim did something recent with this on another PR.  I will take a look at this to see if I can reduce this memory consumption.

2. I renamed the change fragment.

3. I am not sure what issues still exist in JWST.  Looking at the JWST PR it appears I have addressed everything.  If there is something more that needs to be addressed, please specifically mark it.

4. I need to do a rebase, then I will check everythign locally, then run a full regression test in github.

melanieclarke reviewed (approved):
All clean up items addressed, regtests are passing.  I think this is ready to merge now.

We will need to merge this one first, update the pyproject.toml in jwst to point to main, and then merge the jwst PR.

tapastro reviewed (approved):
👏
